vectordraw/grader.py

```diff
@@ -0,0 +1,345 @@
+    points = check.get('expected')
+    for point in points:
+        point = Point(point[0], point[1])
+        if _dist_line_point(line, point) > tolerance:
+            return _errmsg(
+                'The line {name} does not pass through the correct points.', check, vectors
+            )
+
+
+def check_point_coords(check, points):
+    """
+    Check if coordinates of point targeted by `check` are correct.
+    """
+    point = points[check['point']]
+    tolerance = check.get('tolerance', 1.0)
+    expected = check.get('expected')
+    dist = math.hypot(expected[0] - point.x, expected[1] - point.y)
+    if dist > tolerance:
+        return _errmsg_point('Point {name} is not at the correct location.', check, point)
+
+
+class Point(object):
+    """ Represents a single point on the vector drawing board. """
+    def __init__(self, x, y):
+        self.x = x
+        self.y = y
+
+
+class Vector(object):
+    """ Represents a single vector on the vector drawing board. """
+    def __init__(self, name, x1, y1, x2, y2):  # pylint: disable=too-many-arguments
+        self.name = name
+        self.tail = Point(x1, y1)
+        self.tip = Point(x2, y2)
+        self.length = math.hypot(x2 - x1, y2 - y1)
+        angle = math.degrees(math.atan2(y2 - y1, x2 - x1))
+        if angle < 0:
+            angle += 360
+        self.angle = angle
+
+    def opposite(self):
+        """
+        Return new vector with tip and tail swapped.
+        """
+        return Vector(self.name, self.tip.x, self.tip.y, self.tail.x, self.tail.y)
+
+
+class Grader(object):
+    """
+    Implements grading logic for student answers to Vector Drawing exercises.
+    """
+    check_registry = {
+        'presence': check_presence,
+        'tail': check_tail,
+        'tip': check_tip,
+        'tail_x': check_tail_x,
+        'tail_y': check_tail_y,
+        'tip_x': check_tip_x,
+        'tip_y': check_tip_y,
+        'coords': check_coords,
+        'length': check_length,
+        'angle': check_angle,
+        'segment_angle': check_segment_angle,
+        'segment_coords': check_segment_coords,
+        'points_on_line': check_points_on_line,
+        'point_coords': check_point_coords,
+    }
+
+    def __init__(self, success_message='Test passed', custom_checks=None):
+        self.success_message = success_message
+        if custom_checks:
+            self.check_registry.update(custom_checks)
+
+    def grade(self, answer):
+        """
+        Check correctness of `answer` by running checks defined for it one by one.
+
+        Short-circuit as soon as a single check fails.
+        """
+        check_data = dict(
+            vectors=self._get_vectors(answer),
+            points=self._get_points(answer),
+        )
+        for check in answer['checks']:
+            check_data['check'] = check
+            check_fn = self.check_registry[check['check']]
+            args = [check_data[arg] for arg in inspect.getargspec(check_fn).args]
```

Comment:
I just noticed this hack I introduced when adding points as a problem type.  It's main purpose was to stay backwards-compatible with existing external graders.  We might now be able to simply pass `check, vectors, points` as arguments to all graders.  (Up to you whether you want to change this.  it would require editing the parameter lists of all functions in this file, and updaing all the tests.)